File: frontend/src/admin/constants/orderStatus.ts
export const ORDER_STATUSES = [
  "CREATED",
  "PAID",
  "PACKED",
  "SHIPPED",
  "DELIVERED",
  "CANCELLED",
] as const;

export type OrderStatus = (typeof ORDER_STATUSES)[number];

export const ORDER_STATUS_TRANSITIONS: Record<
  OrderStatus,
  OrderStatus[]
> = {
  CREATED: ["PAID", "CANCELLED"],
  PAID: ["PACKED", "CANCELLED"],
  PACKED: ["SHIPPED"],
  SHIPPED: ["DELIVERED"],
  DELIVERED: [],
  CANCELLED: [],
};

File: backend/src/main/java/com/buildwithrani/backend/order/enums/OrderStatus.java
package com.buildwithrani.backend.order.enums;

import java.util.EnumSet;
import java.util.Set;

public enum OrderStatus {

    CREATED {
        @Override
        public Set<OrderStatus> allowedNext() {
            return EnumSet.of(PAID, CANCELLED);
        }
    },

    PAID {
        @Override
        public Set<OrderStatus> allowedNext() {
            return EnumSet.of(PACKED, CANCELLED);
        }
    },

    PACKED {
        @Override
        public Set<OrderStatus> allowedNext() {
            return EnumSet.of(SHIPPED);
        }
    },

    SHIPPED {
        @Override
        public Set<OrderStatus> allowedNext() {
            return EnumSet.of(DELIVERED);
        }
    },

    DELIVERED {
        @Override
        public Set<OrderStatus> allowedNext() {
            return EnumSet.noneOf(OrderStatus.class);
        }
    },

    CANCELLED {
        @Override
        public Set<OrderStatus> allowedNext() {
            return EnumSet.noneOf(OrderStatus.class);
        }
    };

    public abstract Set<OrderStatus> allowedNext();

    public boolean canTransitionTo(OrderStatus nextStatus) {
        return allowedNext().contains(nextStatus);
    }

    public boolean isTerminal() {
        return this == DELIVERED || this == CANCELLED;
    }
}
